frame/composable-traits/src/bonded_finance.rs

```diff
@@ -0,0 +1,82 @@
+use frame_support::pallet_prelude::*;
+use scale_info::TypeInfo;
+use sp_runtime::{traits::Zero, ArithmeticError};
+
+use crate::math::SafeArithmetic;
+
+pub trait BondedFinance {
+	type AccountId;
+	type AssetId;
+	type Balance;
+	type BlockNumber;
+	type BondOfferId;
+
+	/// Create a new offer.
+	fn offer(
+		from: &Self::AccountId,
+		offer: BondOffer<Self::AssetId, Self::Balance, Self::BlockNumber>,
+	) -> Result<Self::BondOfferId, DispatchError>;
+
+	/// Bond for an offer.
+	fn bond(
+		offer: Self::BondOfferId,
+		from: &Self::AccountId,
+		contracts: Self::Balance,
+	) -> Result<Self::Balance, DispatchError>;
+}
+
+/// The Bond duration.
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub enum BondDuration<BlockNumber> {
+	/// Finite duration, liquidity is returned after a number of `blocks`.
+	Finite { blocks: BlockNumber },
+	/// Infinite duration, the protocol is now owning the liquidity
+	Infinite,
+}
+
+/// The Bond offer.
+#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct BondOffer<AssetId, Balance, BlockNumber> {
+	/// Asset to be locked. Unlockable after `duration`.
+	pub asset: AssetId,
+	/// Price of a bond.
+	pub price: Balance,
+	/// Number of bonds. We use the Balance type for the sake of simplicity.
+	pub contracts: Balance,
+	/// Duration for which the asset has to be locked.
+	pub duration: BondDuration<BlockNumber>,
+	/// Asset given as reward.
+	pub reward_asset: AssetId,
+	/// Total reward.
+	pub reward_amount: Balance,
+	/// Duration after which the reward can be claimed.
+	pub reward_duration: BlockNumber,
+}
+
+impl<AssetId, Balance: Zero + PartialOrd + SafeArithmetic, BlockNumber: Zero>
+	BondOffer<AssetId, Balance, BlockNumber>
+{
+	pub fn completed(&self) -> bool {
+		self.contracts.is_zero()
+	}
+	pub fn total_price(&self) -> Result<Balance, ArithmeticError> {
+		self.contracts.safe_mul(&self.price)
+	}
+	pub fn valid(&self, min_transfer: Balance, min_reward: Balance) -> bool {
+		let valid_duration = match &self.duration {
+			BondDuration::Finite { blocks } => !blocks.is_zero(),
```

Comment:
NonZero block number :)

Reply:
NonZero is not pragmatic to use, the thing is that you don't have a generic NonZero unless I missed something. you have to chose NonZeroU8, NonZeroU16 etc...